Add offline search functionality for addresses

pac-guerreiro opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

Add offline search functionality for addresses

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/30123
PROPOSAL: https://github.com/Expensify/App/issues/30123#issuecomment-1773942379


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Go to staging.new.expensify.com
2. Click on FAB> Submit Expense > Distance tab
3. Select different locations several times
4. Go offline
5. Select Start point
6. Enter any letter into search field and check if search result will update according to entered letter

Since this PR introduces changes to `FormWrapper` which is used in 80 files, it’s important to test that these changes don’t break other forms.

The following are some tests cases of other forms:

- Assign Task
    1. On home, click on ( + ) FAB and select “Assign task”
    2. Confirm that the form is rendered correctly

- Start chat > Room
    1. On home, click on ( + ) FAB and select “Start chat”
    2. Select “Room” tab
    3. Confirm that the form is rendered correctly

- Track expense > Confirmation
    1. On home, click on ( + ) FAB and select “Track expense”
    2. Select “Manual” tab
    3. Input any amount then click on “Next”
    4. Click on “Description”
    5. Confirm that the form is rendered correctly then go back
    6. Click on “Merchant”
    7. Confirm that the form is rendered correctly then go back
    8. Click on “Date”
    9. Confirm that the form is rendered correctly

- Settings > Profile
    1. On home, click on “Settings”
    2. On settings, click on “Profile”
    3. On profile, click on “Display name”
    4. Confirm that the form is rendered correctly then go back
    5. On profile, click on “Status”
    6. Confirm that the form is rendered correctly then go back
    7. On profile, click on “Legal name”
    8. Confirm that the form is rendered correctly then go back
    9. On profile, click on “Date of birth”
    10. Confirm that the form is rendered correctly then go back
    11. On profile, click on “Address”
    12. Confirm that the form is rendered correctly

- Settings > Security > Close account
    1. On home, click on “Settings”
    2. On settings, click on “Security”
    3. On security, click on “Close account”
    4. Confirm that the form is rendered correctly

- Workspace > Profile > Description
    1. On home, click on “Settings”
    2. On settings, click on “Workspaces”
    3. On workspaces, click on any workspace in which you are an admin or create one
    4. On workspace home, click on “Profile”
    5. On workspace profile, click on “Description”
    6. Confirm that the form is rendered correctly

- Workspace > Members > Invite Member > Invite Message
    1. On home, click on “Settings”
    2. On settings, click on “Workspaces”
    3. On workspaces, click on any workspace in which you are an admin or create one
    4. On workspace home, click on “Members”
    5. On workspace members, click on “Invite member”
    6. Select any user then click on “Next”
    7. Confirm that the form is rendered correctly

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as above

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as above

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/5f2d417d-b519-42cf-b4aa-6eec88fd72c6

https://github.com/Expensify/App/assets/48553277/bd987fdf-33f5-49c2-9c9e-a9a7e131ee71

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/dcc3cdcd-5ee4-4faf-ba5a-f1741354ad39

https://github.com/Expensify/App/assets/48553277/1569f4dd-c937-4361-a0ac-f939cbaecc07

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/7c66b280-735e-4348-8581-335247193cf0

https://github.com/Expensify/App/assets/48553277/6dd47dec-e424-4a1c-bd75-8b47dc1f25d4

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/8250a6e0-c730-45dc-bcf7-eff537edea00

https://github.com/Expensify/App/assets/48553277/2d6cfa07-15db-4cac-b04b-7c0b0397665d

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/57c0142f-2893-4151-9efe-ef295abe5a09

https://github.com/Expensify/App/assets/48553277/8630d865-5f1f-4e00-8aa2-b2ab3f3c1991

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
https://github.com/Expensify/App/assets/48553277/e7f776f9-1ecb-48dc-8814-d4276d848a03

https://github.com/Expensify/App/assets/48553277/e4ec354d-74e6-4ec8-9bea-5423a79a67d0

</details>

cooldev900 committed:
offline search functinality for addresses

cooldev900 committed:
eliminate lint error

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
eliminate extra spaces

cooldev900 committed:
eliminate lint error

cooldev900 committed:
resolve conflict

cooldev900 committed:
resolve conflict

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
show hint in offline state

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
restore hint and update isSearchedPlace funtion

cooldev900 committed:
update function name and jsDoc

cooldev900 committed:
handle overflow

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
fix prettier error

cooldev900 committed:
add space on the comment

cooldev900 committed:
restore the update by merge

cooldev900 committed:
add document for place

cooldev900 committed:
add the document for geometry

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
fix overflow on ios

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
merge main

cooldev900 committed:
upadte height and back ground color

cooldev900 committed:
calculate list view style

cooldev900 committed:
update style depending device

cooldev900 committed:
fix lint error

cooldev900 committed:
fix prettier error

cooldev900 committed:
create native file for mobile

cooldev900 committed:
create getListViewHeight function

cooldev900 committed:
resolve conflict

cooldev900 committed:
fix typo error

cooldev900 committed:
resolve conflict

cooldev900 committed:
fix lint

cooldev900 committed:
add const and fix bug

cooldev900 committed:
remove lodash

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
fix styles

cooldev900 committed:
Merge branch 'main' into issues/30123

cooldev900 committed:
fix button movement error

cooldev900 committed:
adjust form height on offline state

cooldev900 committed:
update height in ios

cooldev900 committed:
fix overflow styles

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
reset focus state

cooldev900 committed:
use windowHeight instead of 100vh

cooldev900 committed:
updating styles for overflow

cooldev900 committed:
ios fix

cooldev900 committed:
fiix overflow issues in ios and safari

cooldev900 committed:
consider ios safari

cooldev900 committed:
eliminate shrinking list by save button

cooldev900 committed:
solve merge conflict

cooldev900 committed:
Merge branch 'main' of https://github.com/cooldev900/App into issues/30123

cooldev900 committed:
fix lint error

cooldev900 committed:
keep responsivness in web

cooldev900 committed:
resolve conflict

cooldev900 committed:
fix typo error

cooldev900 committed:
update explanation

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/AddressSearch/index.tsx
#	src/components/Form/FormProvider.js
#	src/components/Form/FormWrapper.js
#	src/pages/iou/WaypointEditor.js

pac-guerreiro committed:
fix(AddressSearch): addresses getting cropped

pac-guerreiro commented:
This PR is a continuation of this one -> https://github.com/Expensify/App/pull/31821

pac-guerreiro commented:
@eVoloshchak @neil-marcellini Just to confirm, should the form submit button always stick to the bottom of the screen in all screens?

I basically made changes so that the submit button is always positioned at the bottom of the page, and the container that holds the inputs is scrollable. If this is fine for the whole app, then there's no problem. If not, I would have to only apply this solution to this case which is perfectly doable!

pac-guerreiro committed:
chore: cleanup unused code

pac-guerreiro committed:
chore: remove unused code

pac-guerreiro committed:
refactor: migrate files to typescript

eVoloshchak reviewed (commented):
@pac-guerreiro, it looks like there is no consistency in how we handle the submit button, on some screens it sticks to the bottom, on others it is hidden if you shrink the window. 

https://github.com/Expensify/App/assets/9059945/f1140c37-fc2e-45dc-90c7-9171c25f43d7


We should keep the current behavior on all screens, including the address screen. The current behavior on address screen is that the sumbit button is hidden, which might have a logical explanation: you need the submit button only in cases where you enter address manually (i.e. it isn't on google maps), which is rare, so displaying the actual list is more important. 
So let's keep the current behavior on all screens (including this particular one)

pac-guerreiro commented:
@eVoloshchak Ok, I'll revert my changes then! Thanks 😄

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/Form/FormProvider.js
#	src/components/Form/FormWrapper.js

pac-guerreiro committed:
chore: revert changes

pac-guerreiro commented:
@eVoloshchak Sorry for the delay! It should be working now! 😄

eVoloshchak commented on src/components/AddressSearch/index.tsx:
```diff
@@ -411,10 +434,18 @@ function AddressSearch(
-                            listView: [StyleUtils.getGoogleListViewStyle(displayListViewBorder), styles.overflowAuto, styles.borderLeft, styles.borderRight, !isFocused && {height: 0}],
+                            listView: [
+                                StyleUtils.getGoogleListViewStyle(displayListViewBorder),
+                                getListViewHeight(),
```
There's no need for this to be a function I think
`getListViewOverflow` file can be renamed to `listViewOverflow` and instead of a function, 
`index.js` file will just return `styles.overflowAuto` and
`index.native.js` will return `styles.overflowHidden`

eVoloshchak commented:
Bug:
The submit button isn't fixed at the bottom of the screen, it floats below the content

<img width="1512" alt="Screenshot 2024-02-01 at 21 12 29" src="https://github.com/Expensify/App/assets/9059945/f4ba7f25-f727-4eaf-9d51-aa871666a728">

It seems like this is the case for many pages, but not all


https://github.com/Expensify/App/assets/9059945/c4ef2ab8-a721-4dd9-b56c-4b6e8549aed7

cc: @pac-guerreiro

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/pages/EditReportFieldDatePage.tsx
#	src/pages/EditReportFieldTextPage.tsx
#	src/pages/settings/Security/CloseAccountPage.tsx
#	src/pages/workspace/WorkspaceOverviewPage.js

pac-guerreiro committed:
chore: revert changes

pac-guerreiro committed:
refactor: apply pull request suggestion

pac-guerreiro commented:
@eVoloshchak Feedback applied and bug should be fixed since I reverted the remaining changes I did which I forgot to remove, sorry 😅

eVoloshchak commented:
@pac-guerreiro, no worries!
Looks like there must have been an error during the merge, the bug with list not being scrollable (and the whole page being scrollable) is back


https://github.com/Expensify/App/assets/9059945/638e6326-783f-47c7-9a38-764cb16cdb95

pac-guerreiro commented:
@eVoloshchak I reverted my changes to the FormWrapper, that's why the bug is back!

It's just not possible to achieve what you want without modifying FormWrapper!

The only alternative I'm seeing is to reapply my fix but make it optional to fix the submit button at the bottom. This way the current behaviour on all pages will remain the same, but for this case I can fix the button at the bottom by passing a prop.

eVoloshchak commented:
> The only alternative I'm seeing is to reapply my fix but make it optional to fix the submit button at the bottom. This way the current behaviour on all pages will remain the same, but for this case I can fix the button at the bottom by passing a prop.

Yeah, this sounds good, let's go with this approach

pac-guerreiro committed:
fix: address search suggestions overflowing

pac-guerreiro commented:
@eVoloshchak It should be fixed now!

eVoloshchak commented on src/components/AddressSearch/index.tsx:
```diff
@@ -411,10 +434,18 @@ function AddressSearch(
+                                styles.borderLeft,
+                                styles.borderRight,
+                                styles.flexGrow0,
+                                !isFocused && {height: 0},
```
```suggestion
                                !isFocused && styles.h0,
```

eVoloshchak commented on src/components/AddressSearch/index.tsx:
```diff
@@ -411,10 +434,18 @@ function AddressSearch(
                             row: [styles.pv4, styles.ph3, styles.overflowAuto],
                             description: [styles.googleSearchText],
                             separator: [styles.googleSearchSeparator],
+                            container: [{maxHeight: '100%'}],
```
```suggestion
                            container: [styles.mh100],
```

eVoloshchak commented on src/components/Form/FormWrapper.tsx:
```diff
@@ -99,7 +99,7 @@ function FormWrapper({
                 onSubmit={onSubmit}
             >
-                {children}
+                <View style={{flex: 1}}>{children}</View>
```
```suggestion
                <View style={styles.flex1}>{children}</View>
```

eVoloshchak reviewed (changes requested):
Looking good, awesome work!
There are a couple of styles that need to be defined inside `utils/sizing.ts`, but otherwise working good

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/AddressSearch/types.ts

pac-guerreiro committed:
refactor: apply pull request suggestions

pac-guerreiro commented:
@eVoloshchak Feedback applied, thank you! 😄

eVoloshchak commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

eVoloshchak commented:
@pac-guerreiro 
- **Screenshots/Videos** section is missing from PR author checklist
- could you pull the latest `main` please?

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/Form/FormWrapper.tsx

eVoloshchak commented:
@pac-guerreiro, could you please add and populate **Screenshots/Videos** section to the PR author checklist? Similar to how it's done in https://github.com/Expensify/App/pull/31821

eVoloshchak commented:
@pac-guerreiro, no, we do this to double-check (by contributor and by C+) that all platforms work as expected. So it's important those videos feature the latest changes from this PR

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak reviewed (commented):
Re-tested all platforms, looks good on Web/Desktop, found an issue on native (iOS/Android):
Save button is floating above the keyboard, which shrinks down the recents list

https://github.com/Expensify/App/assets/9059945/818c11f7-8e33-4519-b68c-bb2e085c53ad

Here's what it should look like, this will make the behavior consistent with mWeb

https://github.com/Expensify/App/assets/9059945/5427a38e-4b74-4b7f-8aa8-1f6ab6e754a7

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
fix: submit button not fixed at the bottom of the screen on both android and ios

pac-guerreiro commented:
@eVoloshchak It should be fixed now! I fixed the submit button to the bottom of the screen!

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak reviewed (commented):
@pac-guerreiro, this looks good on native, but breaks the button positioning when resizing a window on web/desktop

https://github.com/Expensify/App/assets/9059945/5f1fb938-adc4-4e05-aaa5-b06fd9d297e3

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/AddressSearch/index.tsx
#	src/components/AddressSearch/types.ts
#	src/pages/iou/request/step/IOURequestStepWaypoint.tsx

pac-guerreiro committed:
fix: submit button positioning on address search

eVoloshchak commented on src/pages/iou/request/step/IOURequestStepWaypoint.tsx:
```diff
@@ -254,10 +262,10 @@ export default withOnyx<IOURequestStepWaypointProps, IOURequestStepWaypointOnyxP
         key: ONYXKEYS.NVP_RECENT_WAYPOINTS,
 
-        // Only grab the most recent 5 waypoints because that's all that is shown in the UI. This also puts them into the format of data
+        // Only grab the most recent waypoints because that's all that is shown in the UI. This also puts them into the format of data
```
```suggestion
        // Only grab the most recent 20 waypoints because that's all that is shown in the UI. This also puts them into the format of data
```

eVoloshchak replied:
Bump on this comment too

pac-guerreiro replied:
@eVoloshchak sorry I missed this comment 😅

eVoloshchak commented on src/components/Form/FormWrapper.tsx:
```diff
@@ -121,26 +121,26 @@ function FormWrapper({
             styles.flex1,
             styles.mh0,
             styles.mt5,
+            children,
```
Why are we changing the order of these props?

eVoloshchak replied:
Bump on this comment

pac-guerreiro replied:
@eVoloshchak I don't recall why I changed this 😅 I'm going to revert this changes, thank you for noticing it!

eVoloshchak commented on src/pages/iou/request/step/IOURequestStepWaypoint.tsx:
```diff
@@ -164,6 +166,11 @@ function IOURequestStepWaypoint({
             testID={IOURequestStepWaypoint.displayName}
+            style={[styles.overflowHidden]}
+            // This keeps the submit button at the bottom of the screen on iOS
+            shouldEnableKeyboardAvoidingView={Platform.OS !== 'ios'}
```
We try to avoid usage of `Platform.OS` directly inside the components, take a look at 
https://github.com/expensify/app?tab=readme-ov-file#platform-specific-file-extensions

> In most cases, the code written for this repo should be platform-independent. In such cases, each module should have a single file, index.js, which defines the module's exports. There are, however, some cases in which a feature is intrinsically tied to the underlying platform. In such cases, the following file extensions can be used to export platform-specific code from a module:
>- Mobile => index.native.js
>- iOS Native App/Android Native App => index.ios.js/index.android.js
>- Web => index.website.js
>- Desktop => index.desktop.js

eVoloshchak commented on src/pages/iou/request/step/IOURequestStepWaypoint.tsx:
```diff
@@ -195,45 +202,46 @@ function IOURequestStepWaypoint({
                     onSubmit={submit}
+                    submitFlexEnabled={false}
+                    // This keeps the list of waypoints above the keyboard on android
+                    submitButtonStyles={keyboardHeight && Platform.OS === 'android' ? {height: keyboardHeight} : null}
```
This calculation doesn't seem to be correct, there is a gap between the list and a keyboard (Pixel 6 simulator, Android 13)

https://github.com/Expensify/App/assets/9059945/59e0435a-d2e4-41f1-875f-bb2b5f620ca8

eVoloshchak commented:
[This](https://github.com/Expensify/App/pull/35045#pullrequestreview-1889107441) bug is now present on mWeb (both Safari and Chrome)


https://github.com/Expensify/App/assets/9059945/a15293e9-3c10-4886-a25f-20d3b7c75faa
<img width="322" alt="Screenshot 2024-02-28 at 21 49 28" src="https://github.com/Expensify/App/assets/9059945/4d9bb7d7-1023-4758-8aa2-eff45451df42">

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak commented:
@pac-guerreiro, gentle bump on https://github.com/Expensify/App/pull/35045#issuecomment-1969895600

pac-guerreiro commented:
@eVoloshchak Sorry I thought I had replied to you but it seems it didn't go through 😔 I'm on vacations until next monday!

Regarding the issues I'm facing with this task, I've been trying different solutions to the problems you've been reporting but it seems like I'm chasing my own tail 😅

Here is what I've tried:

- Setting the container min height as the window height but while the submit button stays on the bottom, the options list can get clipped by the keyboard on mobile native and web;

- So I tried detecting the keyboard height and apply a calculation to maintain the spacing between the keyboard and the options list, but on mobile web I'm not able to detect the keyboard height when the keyboard is shown because react-native-web mocks the keyboard api;

- Then I tried hiding the submit button and let the container height change has the window changes, but there can be some moments where the user will see the button showing for like a second before disappearing, and besides that on desktop native and web hidding the button will be weird.

So, from what I saw in the code, previously we'd only show 5 location options on the list and now we're showing 20, thus the list increased in size. What I propose here is that we set the size of the list to always show a maximum of 5 options at a time and let the user scroll the list to access the remaining options. What do you think?

I'm open to other ideas if you have any! 😄

eVoloshchak commented:
> So, from what I saw in the code, previously we'd only show 5 location options on the list and now we're showing 20, thus the list increased in size. What I propose here is that we set the size of the list to always show a maximum of 5 options at a time and let the user scroll the list to access the remaining options. What do you think?

Let's resort to this if we exhaust other options, increasing the amount of places was one of two goals of this PR, I think we should be able to do it

> Setting the container min height as the window height but while the submit button stays on the bottom, the options list can get clipped by the keyboard on mobile native and web

Could you explain what you mean by 'clipped'? Is it the blank line above the keyboard like on the video below?


https://github.com/Expensify/App/assets/9059945/d6e4a7dd-7880-4181-af5e-3e7ed6fb889b

This is how it currently looks on Search page on iOS, I think this is a separate bug

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/Form/FormWrapper.tsx

pac-guerreiro commented:
@eVoloshchak
Sorry for the delay, basically if I maintain the height of the page when the virtual keyboard shows up on iOS, then half the list becomes unreachable because it's behind the virtual keyboard!

pac-guerreiro commented:
@eVoloshchak What should I do? 😄

eVoloshchak commented:
> Sorry for the delay, basically if I maintain the height of the page when the virtual keyboard shows up on iOS, then half the list becomes unreachable because it's behind the virtual keyboard! Which seems similar to what you recorded on the search page

@pac-guerreiro, could you show a screenshot of that please?
If we're facing the same issue as on the search page, than I think we can proceed and tackle this issue separately (if it isn't being explored already)

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/CONST.ts
#	src/pages/iou/request/step/IOURequestStepWaypoint.tsx

pac-guerreiro commented:
@eVoloshchak 

Here is the observed behaviour on iOS:

https://github.com/Expensify/App/assets/48553277/ea41fe06-0de5-4fbc-9d29-5b442aa4c05d


Also, I managed to fix the extra bottom gap between the list and the virtual keyboard on android:

![Screenshot_1710950005](https://github.com/Expensify/App/assets/48553277/1800f1c8-40e1-4f48-b80c-623a04148790)

pac-guerreiro commented:
@eVoloshchak 

I just tweaked my last solution a bit, what do you think of this behaviour?

https://github.com/Expensify/App/assets/48553277/4cc299a9-3412-4c3f-aa91-d293ee106757

pac-guerreiro committed:
fix: address suggestions list sizing

eVoloshchak commented:
@pac-guerreiro, this breaks web/Desktop, the button is hidden when the input is focused and you have to press on the item twice to select it

https://github.com/Expensify/App/assets/9059945/0fc62762-5c54-42a3-9336-a43adb37edb0

pac-guerreiro commented:
@eVoloshchak Ah crap, I can solve that by preventing the logic from running on non mobile platforms, sounds good? 😄

pac-guerreiro commented:
@eVoloshchak I just remembered that on mobile chrome and safari the `Platform.OS === 'web`', so I can't ignore this logic on web 😞

pac-guerreiro committed:
fix: submit button visibility on all platforms

pac-guerreiro commented:
@eVoloshchak 

I just worked on a solution that seems to work on all platforms! Also, I extracted the solution into a separate hook which can be useful in other use cases!

Let me know if you find other issues related to this!

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/AddressSearch/types.ts

dubielzyk-expensify reviewed (dismissed):
From a visual perspective we're looking good here 😄

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,66 @@
+    // otherwise it's hidden
+    useEffect(() => {
+        const dimensionsListener = Dimensions.addEventListener('change', ({window}) => {
+            if (Platform.OS !== 'web') {
```
```suggestion
            if (getPlatform() === CONST.PLATFORM.WEB) {
```

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,66 @@
+    // Web: the submit button is only shown when the window height is the same or greater,
+    // so executing this function won't do anything
+    const showSubmitButton = () => {
+        if (!['ios', 'android'].includes(Platform.OS)) {
```
```suggestion
    const platform = getPlatform();
    const isNative = platform === CONST.PLATFORM.IOS || platform === CONST.PLATFORM.ANDROID;
        if (!isNative) {
```

eVoloshchak commented:
@pac-guerreiro, awesome, we're getting very close, this looks good on mobile devices (both mWeb and native)
There is one small bug on desktop and Web though: if you resize the window (reduce its height), the submit button is hidden

https://github.com/Expensify/App/assets/9059945/9341c039-588c-4715-86f9-90cffc718c09

eVoloshchak commented on src/components/Form/FormWrapper.tsx:
```diff
@@ -63,7 +63,7 @@ function FormWrapper({
     const styles = useThemeStyles();
     const formRef = useRef<RNScrollView>(null);
-    const formContentRef = useRef<View>(null);
+    const formContentRef = useRef<RNView>(null);
```
I think I'm missing something, what is the purpose of this change? 
Aren't `RNView` and `View` the same?

pac-guerreiro replied:
Yes they're the same, I'm going to revert this!

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,66 @@
+import useThemeStyles from './useThemeStyles';
+import useWindowDimensions from './useWindowDimensions';
+
+// Usefull when there's need to hide submit button, from FormProvider,
```
```suggestion
// Useful when there's a need to hide the submit button from FormProvider,
```

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,66 @@
+    };
+
+    // When the submit button is hidden there's need to manually
+    // add it's bottom style to the FormProvider style prop,
```
```suggestion
    // add its bottom style to the FormProvider style prop,
```

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
refactor: apply pull request suggestions

pac-guerreiro commented:
> @pac-guerreiro, awesome, we're getting very close, this looks good on mobile devices (both mWeb and native) There is one small bug on desktop and Web though: if you resize the window (reduce its height), the submit button is hidden
> 
>  Screen.Recording.2024-03-25.at.10.02.25.mov

@eVoloshchak regarding this issue, I noticed that there are some functions in the code base that check if the app is running on desktop web browser but I'm not sure how safe they are 😅

Otherwise I think I could only apply the logic to hide the submit button on web if the window height > window width, which means that the device is running on portrait which is the most common case for mobile web.

Or I could define a minimum height for the submit button to be displayed, but that could also miss some edge cases 😅

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak commented:
> Otherwise I think I could only apply the logic to hide the submit button on web if the window height > window width, which means that the device is running on portrait which is the most common case for mobile web.

@pac-guerreiro, yeah, let's do that, I think we already use `isSmallScreenWidth` from `useWindowDimensions` to toggle between mobile/desktop interface

pac-guerreiro committed:
fix: submit button disappearing when web/desktop window is resized

pac-guerreiro committed:
refactor: apply pull request suggestions

pac-guerreiro commented:
@eVoloshchak Thank you for the suggestion! It works 🥳 

Let me know if you find any edge case 😄

eVoloshchak reviewed (changes requested):
All platforms are looking good from the UI perspective, awesome!
There's a single bug left I think:

1. Click on FAB> Request money> Distance tab
2. Go offline
3. Ented an address that isn't saved (random jibberish)
4. App is correctly displaying "No results found"
5. Click on save
6. Notice the point is saved, even though we haven't selected a valid address (It should behave the same as online, the only difference is instead of "Please select a suggested address or use current location" we should display "Please select a suggested address")

https://github.com/Expensify/App/assets/9059945/d22104fe-9c8b-4bc3-be22-6ae1b903962f

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,73 @@
+    const initialWindowHeightRef = useRef(windowHeight);
+    const isSmallScreenWidthRef = useRef(isSmallScreenWidth);
+    const {bottom} = useSafeAreaInsets();
+
```
We can move the following two lines here to avoid duplication
```js
const platform = getPlatform();
const isNative = platform === CONST.PLATFORM.IOS || platform === CONST.PLATFORM.ANDROID;
```

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,73 @@
+    const {windowHeight, isSmallScreenWidth} = useWindowDimensions();
+    const [isSubmitButtonVisible, setIsSubmitButtonVisible] = useState(true);
+    const initialWindowHeightRef = useRef(windowHeight);
+    const isSmallScreenWidthRef = useRef(isSmallScreenWidth);
```
I think I'm missing something, could you elaborate on why we need initial values (why do we use `useRef`?)

pac-guerreiro replied:
@eVoloshchak useRef is used here to store the initial value of window height, so that it can be compared with the current window height value to determine if the virtual keyboard is shown or not. About isSmallScreenWidthRef I had to use a ref too because the callback I pass to the `Dimensions.addEventListener` cannot have dependencies, otherwise multiple event listeners would be set 😄

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
refactor: apply pull request suggestions

pac-guerreiro commented:
@eVoloshchak Thank you for your feedback! I applied your suggestions, let me know if there's anything else 😄

shawnborton reviewed (dismissed):
All good from a design perspective

pac-guerreiro commented:
@eVoloshchak kind bump 😄

eVoloshchak commented:
@pac-guerreiro, take a look at item 5 in https://github.com/Expensify/App/pull/35045#pullrequestreview-1973965585, we shouldn't show the "Please select a suggested address or use current location" error, since it's not possible;e to use current location while offline

<img width="374" alt="Screenshot 2024-04-04 at 20 25 07" src="https://github.com/Expensify/App/assets/9059945/72519261-af11-40ab-bb8c-8057ad18e8f9">

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
fix: wrong IOU distance request address error message when offline

pac-guerreiro commented:
@eVoloshchak Oh sorry I missed that part! 😅 Now it's done! 🥳

pac-guerreiro committed:
fix: duplicate sizing name

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/pages/iou/request/step/IOURequestStepWaypoint.tsx

pac-guerreiro committed:
refactor(prettier): code style

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro commented:
@eVoloshchak I resolved conflicts today, feel free to give this a spin again when you have the time 😄

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/components/AddressSearch/index.tsx
#	src/components/AddressSearch/types.ts

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,71 @@
+import useThemeStyles from './useThemeStyles';
+import useWindowDimensions from './useWindowDimensions';
+
+// Useful when there's a need to hide the submit button, from FormProvider,
```
```suggestion
// Useful when there's a need to hide the submit button from FormProvider,
```

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,71 @@
+    // otherwise it's hidden
+    useEffect(() => {
+        const dimensionsListener = Dimensions.addEventListener('change', ({window}) => {
+            if (getPlatform() !== CONST.PLATFORM.WEB || !isSmallScreenWidthRef.current) {
```
```suggestion
            if (platform !== CONST.PLATFORM.WEB || !isSmallScreenWidthRef.current) {
```

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,71 @@
+        setIsSubmitButtonVisible(false);
+    };
+
+    // When the submit button is hidden there's need to manually
```
```suggestion
    // When the submit button is hidden there's a need to manually
```

eVoloshchak reviewed (changes requested):
Last finishing touches, thank you for the patience!

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
refactor: apply pull request suggestions

pac-guerreiro commented:
@eVoloshchak 

Yesterday I applied your feedback but I forgot to tell you here, sorry!

eVoloshchak commented:
> Yesterday I applied your feedback but I forgot to tell you here, sorry!

No worries! The most reliable way is to re-request a review from C+, that way the C+ is notified and the PR is displayed in "Needs review" section

eVoloshchak reviewed (commented):
Bug:
The second line in the list item is not displayed (it contained the country before)
This means that, for example, Lichtenstein is blank, since it has only the second line
![Screenshot_20240415-155240](https://github.com/Expensify/App/assets/9059945/fba4f557-7b5f-48b2-a610-7ce0b3464bf6)

Here's what it should like:

![Screenshot_20240415-154515](https://github.com/Expensify/App/assets/9059945/bb6674da-1311-4bcf-996b-7cb43de34184)

This is Android only, not sure if it's device-specific, but I've tested on OnePlus6 running Android 11

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
fix: address search titles missing on mobile

pac-guerreiro committed:
Merge branch 'main' into issues/30123

eVoloshchak reviewed (commented):
Bug:
The submit button is positioned too close to the bottom of the page (with no padding). It's reproducible only when there's a couple of items in the list already

<img width="861" alt="Screenshot 2024-04-16 at 19 54 10" src="https://github.com/Expensify/App/assets/9059945/da91c7fa-3f07-4068-85bf-f72b307581e7">


![Simulator Screenshot - iPhone 15 Pro - 2024-04-16 at 19 54 00](https://github.com/Expensify/App/assets/9059945/7117d767-45d5-489c-a08e-4729e10d89f6)

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro commented:
@eVoloshchak 

This bug was caused by https://github.com/Expensify/App/pull/38039. I just left a comment on it and will discuss with them what should be the next steps 😄

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro commented:
@eVoloshchak 

Seems like a fix was already merged - https://github.com/Expensify/App/pull/40473 - so can you re-test this? 😄

eVoloshchak reviewed (approved):
Looks good and tests well!

pac-guerreiro commented:
@eVoloshchak 

PR description updated accordingly 😄

pac-guerreiro commented:
Bump @neil-marcellini 😄

neil-marcellini commented:
Ah great it's finally ready! I'll review today.

neil-marcellini commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,72 @@
+    const initialWindowHeightRef = useRef(windowHeight);
+    const isSmallScreenWidthRef = useRef(isSmallScreenWidth);
+    const {bottom} = useSafeAreaInsets();
+    const platform = getPlatform();
+    const isNative = platform === CONST.PLATFORM.IOS || platform === CONST.PLATFORM.ANDROID;
```
Can we somehow do this with platform specific files? That's what our guidelines are.

https://github.com/Expensify/App/blob/db01cf5b4aa52069cf416d7a1a84ea3bdbfeea8f/README.md?plain=1#L480

neil-marcellini replied:
I'm pretty sure we can do it, since I already see some hooks with platforms specific files.

neil-marcellini commented on src/pages/iou/request/step/IOURequestStepWaypoint.tsx:
```diff
@@ -93,10 +94,8 @@ function IOURequestStepWaypoint({
-        // If the user is online, and they are trying to save a value without using the autocomplete, show an error message instructing them to use a selected address instead.
-        // That enables us to save the address with coordinates when it is selected
-        if (!isOffline && waypointValue !== '' && waypointAddress !== waypointValue) {
-            ErrorUtils.addErrorMessage(errors, `waypoint${pageIndex}`, 'distance.errors.selectSuggestedAddress');
+        if (waypointValue !== '' && waypointAddress !== waypointValue) {
+            ErrorUtils.addErrorMessage(errors, `waypoint${pageIndex}`, `distance.errors.${isOffline ? 'selectSuggestedAddress' : 'selectSuggestedAddressOrCurrentLocation'}`);
         }
```
We still want to allow users to enter an address manually outside of the pre-defined waypoints while offline right? Or are we now preventing that to ensure that it's a valid address?

I don't remember us discussing this, maybe I forgot, but if we go this route I want to bring it up with the team more broadly. It could be a helpful solution the the fact that we don't handle it well when an invalid address is entered while offline. However, it also seems overly limiting and goes against our offline first philosophy that you should be able to do as much as possible while offline, so I think we should remove the restriction.

eVoloshchak replied:
> We still want to allow users to enter an address manually outside of the pre-defined waypoints while offline right?

I asked for this change since we don't allow it offline, but I wasn't aware that we allow an arbitrary address while offline. 
"It could be a helpful solution the the fact that we don't handle it well when an invalid address is entered while offline. However, it also seems overly limiting and goes against our offline first philosophy that you should be able to do as much as possible while offline, so I think we should remove the restriction." That makes total sense, I agree we should remove this restriction. 
Apologies for the confusion everyone, @pac-guerreiro, could you roll back this change please?

pac-guerreiro replied:
@neil-marcellini @eVoloshchak 

Change is now reverted, thank you both for the help! 😄

neil-marcellini commented on src/components/Form/FormWrapper.tsx:
```diff
@@ -104,7 +104,7 @@ function FormWrapper({
                 style={[style, safeAreaPaddingBottomStyle.paddingBottom ? safeAreaPaddingBottomStyle : styles.pb5]}
             >
-                {children}
+                <View style={styles.flex1}>{children}</View>
```
Did you test that this works with all existing forms in the app? Or is it unlikely to break anything?

neil-marcellini commented on src/components/Form/FormWrapper.tsx:
Since there's some modifications here let's make sure we thoroughly test all the forms.

neil-marcellini reviewed (changes requested):
Overall looking really good. Great work! There are some changes needed based on some of our guidelines, and I would like to see some more testing. I'll check back tomorrow.

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
refactor: revert change that prevents user from writing address in offline mode

neil-marcellini reviewed (changes requested):
Your latest changes look good, but we still need to update the hook to use platforms specific files.

Also, please include some test steps to check that other forms in the app are working, and please include videos of those tests along with @eVoloshchak. You don't have to test every single form, but a testing a couple important ones is a good idea.

pac-guerreiro commented:
@neil-marcellini 

Sorry for the delay in response! Unfortunately I've been having other tasks in my plate which require my focus, but I'll provide the necessary videos as soon as possible! 

Thank you for your patience 🙏

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/languages/en.ts
#	src/languages/es.ts
#	src/pages/iou/request/step/IOURequestStepWaypoint.tsx

pac-guerreiro committed:
refactor: split platform specific login into separate files

eVoloshchak commented:
Not sure why the TS check is failing, [doesn't look](https://github.com/Expensify/App/actions/runs/9110247829/job/25044869792) like it's related to this PR, @pac-guerreiro, could you pull `main` one more time please?

eVoloshchak commented on src/hooks/useSubmitButtonVisibility.ts:
```diff
@@ -0,0 +1,58 @@
+    // When the submit button is hidden there's a need to manually
+    // add its bottom style to the FormProvider style prop,
+    // otherwise the form content will touch the bottom of the page/screen
+    const formStyle = !isSubmitButtonVisible && bottom === 0 && styles.mb5;
```
This is duplicated in both files, we can move this outside of `useSubmitButtonVisibility` by using it directly in `IOURequestStepWaypoint`

pac-guerreiro replied:
@eVoloshchak 

Good point, but I think it would be better for this style to be usable across components. I could extract this into a separate function that receives isSubmitButtonVisible as an argument and returns the style, what do you think?

eVoloshchak commented on src/pages/iou/request/step/IOURequestStepWaypoint.tsx:
```diff
@@ -186,45 +188,47 @@ function IOURequestStepWaypoint({
                 />
                 <FormProvider
-                    style={[styles.flexGrow1, styles.mh5]}
+                    style={[styles.flex1, styles.mh5, formStyle]}
```
```suggestion
                    style={[styles.flex1, styles.mh5, !isSubmitButtonVisible && bottom === 0 && styles.mb5]}
```

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
Merge branch 'main' into issues/30123

pac-guerreiro committed:
fix: submit button overlapping form content

pac-guerreiro commented:
@neil-marcellini @eVoloshchak 

It seems I missed an edge case that affected other forms, in which the submit button would overlap the form content when the virtual keyboard was shown. I fixed it!

I also added additional testing steps and screen recordings of me executing these tests on all platforms 😄 

Thank you for your patience!

neil-marcellini reviewed (dismissed):
The latest batch of changes looks good. Thanks for adding all the extra tests!

neil-marcellini commented:
@eVoloshchak all you for testing now. Please let us know if you find any problems or if it's ready to merge.

eVoloshchak reviewed (approved):
Re-tested the offline search functionality and various Forms, attached the additional videos, tests well
LGTM!

neil-marcellini commented:
Good to go, but we have to wait until the merge freeze is over. Probably only a few days more.

pac-guerreiro commented:
@neil-marcellini 

I think the merge freeze has been lifted 😄

eVoloshchak commented:
@pac-guerreiro, could you please merge the latest main? There are a lot of changes since merge freeze is over, let's re-test this one more time

pac-guerreiro committed:
Merge branch 'main' into issues/30123

# Conflicts:
#	src/CONST.ts

pac-guerreiro commented:
@eVoloshchak 

I merged main and I'll be re-testing this 😄

pac-guerreiro commented:
@eVoloshchak 

Things look the same as in the screen recordings 😄

neil-marcellini commented:
The merge freeze isn't fully over yet, we're merging only low risk PRs, ones with minimal commits and lines changed. This is definitely not one of those. I'll try to let you know when it's over.

neil-marcellini commented:
The merge freeze is over, so finally merging this!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9283356335) to staging by https://github.com/neil-marcellini in version: 1.4.77-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

luacmartins commented:
FYI this change caused regressions:

- https://github.com/Expensify/App/issues/42767
- https://github.com/Expensify/App/issues/42768

arosiclair commented:
Yeah this is being reverted [here](https://github.com/Expensify/App/pull/42777)

luacmartins commented:
@ShridharGoel it seems like we have more blockers stemming from this PR, so I think reverting and letting the original author work on a fix is the right path here.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9309257396) to production by https://github.com/puneetlath in version: 1.4.77-11 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9309047748) to staging by https://github.com/neil-marcellini in version: 1.4.77-11 🚀

platform | result
---|---
🤖 android 🤖|failure ❌
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9309047748) to staging by https://github.com/neil-marcellini in version: 1.4.77-11 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9309257396) to production by https://github.com/puneetlath in version: 1.4.77-11 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/9368314546) to production by https://github.com/mountiny in version: 1.4.78-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅
